core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/variables/runtime/VariableContext.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ *    This file is part of ReadonlyREST.
+ *
+ *    ReadonlyREST is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    ReadonlyREST is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+package tech.beshu.ror.accesscontrol.blocks.variables.runtime
+
+import cats.data.NonEmptyList
+import tech.beshu.ror.accesscontrol.blocks.rules.Rule
+import tech.beshu.ror.accesscontrol.blocks.variables.runtime.MultiExtractable.SingleExtractableWrapper
+
+object VariableContext {
+
+  sealed trait VariableType
+
+  object VariableType {
+    trait User extends VariableType
+    trait CurrentGroup extends VariableType
+    trait Header extends VariableType
+    trait Jwt extends VariableType
+  }
+
+  trait UsingVariable {
```

Comment:
IMO we should not extend rules. AFAIU we use it to provide proper validation of used variables in context of given block. This validation is correlated with configuration loading. All code we can find in rule should be related to processing the ES request. So, it doesn't fit there.

But we can solve this in pretty simple way - by defining proper typeclasses for given rules (at the level of rule instance creation). The validation will take advantage of code from the typeclasses, not rule classes directly.